text/0003-data-frames.md

```diff
@@ -0,0 +1,252 @@
+- Feature Name: data_frames
+- Start Date: 2020-08-05
+- RFC PR: [nushell/rfcs#0003](https://github.com/nushell/rfcs/pull/3)
+- Nushell Issue: [nushell/nushell#0000](https://github.com/nushell/nushell/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+This RFC merges the Row and Table Value types into a single new value type: Frame. Data frames take inspiration from data processing systems like R and Pandas. Data frames will play the fundamental role of modelling data in Nu and will have enough descriptive power to describe all forms of structure, including streaming tables, lists, and objects.
+
+# Motivation
+
+[motivation]: #motivation
+
+The current system has a few unexpected limitations:
+
+- The top-level rows represent a table of rows, but it's unclear how to represent a top-level list of strings vs a stream of strings.
+- A similar ambiguity exists between an "object" (a data structure denoted by key/value pairs) and a table of one row
+- Inner-tables are modelled differently than top-level tables, leading to confusion
+- There is no way to currently represent a matrix
+- As rows are streamed instead of tables, it's not possible to predict how to display this information. This is currently mitigated by buffering some number of rows and treating them as one "table"
+- Likewise, since rows are streamed instead of tables, it's unclear what the user should expect if they request a column that is not present, as this column may appear in the following row instead.
+- When table data is serialized, there is a large amount of duplication, as columns are repeated with each row sent.
+- Additionally, there is currently no way to represent rows using a row literal. We propose a frame literal that will fill this role.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Data frame representation:
+
+```rust
+struct DataFrame {
+    headers: Option<Vec<String>>,
+    rows: Vec<Vec<Value>>,
+    partial_frame_id: Option<Uuid>,
+    is_object: bool,
+}
+```
+
+## Tables
+
+A self-contained table would look like this:
+
+```rust
+let frame = DataFrame {
+    headers: Some(vec!["name", "age"]),
+    rows: vec![
+        vec![Value::from("Bob"), Value::from(30)],
+        vec![Value::from("Sally"), Value::from(43)]
+    ],
+    partial_frame_id: None,
+    is_object: false,
+};
+```
+
+The above code could be created using this Nu syntax:
+
+```sh
+[name: [Bob, Sally], age: [30, 43]]
```

Comment:
How would you express this in nu syntax, and/or the above in rust syntax, if you were to state the shape without data? For instance, to say that Windows `ls` will always have 4 columns and `n` rows and Windows `ls -l` will always have 8 columns and `n` rows.

Alternatively, is there a constructor that says this df is 5 rows and 6 columns?

At some point, I expect we'll have variables that can hold a dataframe. It's hard for me to visualize how this will work in a streaming environment where things are built up and torn down in a pipeline.

Reply:
@fdncred - for the first question, I think you're asking "how do you write types in Nu?" We'll probably need a separate RFC for that, as types will be their own topic.

Or may you're asking how we handle matrices and how this differs from a list of lists?

Reply:
I was asking about initializing a dataframe with a predetermined shape as `ls` may have. `ls -l` on Windows will have a predetermined amount of columns. 

One could think of making dataframes with 2 columns and 3 rows as an empty dataframe except with column names, and then, as the pipeline progresses, update the information in those rows. In order to do this, some type of initialization of the df would have to take place. Maybe the term is dataframe literal. I think this is what you've created here `[name: [Bob, Sally], age: [30, 43]]` but this one is fully populated. Can I do `[name: [], age:[]]` and then populate it later in the pipeline?

Reply:
@fdncred - ah, I think I got it.

There isn't a way to fill in a dataframe, though we could think of creating some API around that like we do for TaggedDictBuilder and related.

Not sure what you mean by populate it later in the pipeline. Since we're passing values through, you'd create a new value. But maybe these helpers would be able to take in a shape and let you fill it in?  Seems doable.

Reply:
Yes, take in a shape and fill it. This may not be exactly `functional` but once we get to scripts i can easily see initializing a dataframe variable (assuming we have variables) and populating it with various pipelines.

1. Define a shape with just columns `df --define columns 3 name size sum`
2. Now populate it `| update sum { ls | get size + accumulator }` (bad syntax but hopefully you get the point)
3. @andrasio probably has examples because he's frequently doing ` | default wassup 0 | blah | blah | update wassup wassam`